src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -115,34 +145,86 @@ internal async Task<IEnumerable<IProjectVersionedValue<RestoreData>>> RestoreAsy
             };
         }
 
-        private async Task<bool> RestoreCoreAsync(ProjectRestoreInfo restoreInfo)
+        private async Task<bool> RestoreCoreAsync(PackageRestoreUnconfiguredInput value)
         {
-            // Restore service always does work regardless of whether the value we pass 
-            // them to actually contains changes, only nominate if there are any.
-            byte[] hash = RestoreHasher.CalculateHash(restoreInfo);
+            ProjectRestoreInfo? restoreInfo = value.RestoreInfo;
+            bool success = false;
 
-            if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
-                return true;
+            try
+            {
+                // Prevent extra restore under some conditions.
+                if (IsProjectConfigurationVersionOutOfDate(value.ConfiguredInputs))
+                {
+                    return false;
+                }
 
-            _latestHash = hash;
+                // Restore service always does work regardless of whether the value we pass 
+                // them to actually contains changes, only nominate if there are any.
+                byte[] hash = RestoreHasher.CalculateHash(restoreInfo!);
 
-            JoinableTask<bool> joinableTask = JoinableFactory.RunAsync(() =>
-            {
-                return NominateForRestoreAsync(restoreInfo, _projectAsynchronousTasksService.UnloadCancellationToken);
-            });
+                if (_latestHash != null && Enumerable.SequenceEqual(hash, _latestHash))
```

Comment:
Minor: This uses Linq which will allocate and be slower than it could be.

```suggestion
                if (_latestHash != null && hash.AsSpan().SequenceEqual(_latestHash))
```